workflows/index-generation/generate_lineage_csvs.py

```diff
@@ -184,24 +206,26 @@ def version_taxon_lineages(
     scheme described above
     """
     previous_lineages = {}
+    previous_lineages_version = None
     if previous_lineages_filename:
         with gzip.open(previous_lineages_filename, "rt") as f:
             for row in csv.DictReader(f):
-                previous_lineages[row["taxid"]] = row
+                previous_lineages[(row["taxid"], row["version_end"])] = row
+                if (
+                    not previous_lineages_version
+                    or previous_lineages_version < row["version_end"]
+                ):
+                    previous_lineages_version = row["version_end"]
 
     with gzip.open(output_filename, "wt") as wf:
         writer = csv.DictWriter(wf, fieldnames=_fieldnames + _versioning_fieldnames)
         writer.writeheader()
 
         with gzip.open(lineages_filename, "rt") as rf:
             for row in csv.DictReader(rf):
-                previous_row = previous_lineages.get(row["taxid"])
-
-                if previous_row:
-                    # If we have a previous row, remove it so we can later
-                    #   iterate through only the rows from previous_lineages
-                    #   that aren't also in lineages
-                    del previous_lineages[row["taxid"]]
+                previous_row = previous_lineages.pop(
+                    (row["taxid"], previous_lineages_version), None
+                )
```

Comment:
To clarify my understanding, the bug we had previously was caused by us only keeping track of the `previous_row` based on `tax_id` and deleting the previous row later on when we compare against the new lineage. This approach didn't allow more than one previous taxon row from being recorded, thus dropping deprecated rows 

This bug fix adds the `version_end` to the `previous_lineages` dictionary in addition to the `tax_id` so that we can keep track of the deprecated rows.

Is my understanding of the issue correct?

Reply:
Yes, that's exactly correct.